File: python3/minimum_depth_of_binary_tree.py
from model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""

class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        res = []
        stack = [(root, 1)]

        while stack:
            node, ls = stack.pop()
            if not node.left and not node.right:
                res.append(ls)
            if node.right:
                stack.append((node.right, ls+1))
            if node.left:
                stack.append((node.left, ls+1))

        return min(res)

File: golang/minimum_depth_of_binary_tree.go
package main

// Space 	: O(n)
// Time	 	: O(n)

import (
	"golang/model"
)

func minDepth(root *model.TreeNode) int {
	level := 1

	if root == nil {
		return 0
	}

	stack := []*model.TreeNode{root}
	for len(stack) > 0 {
		temp := []*model.TreeNode{}

		for _, node := range stack {
			if node.Left == nil && node.Right == nil {
				return level
			}

			if node.Left != nil {
				temp = append(temp, node.Left)
			}

			if node.Right != nil {
				temp = append(temp, node.Right)
			}
		}
		stack = temp
		level++
	}

	return level
}
